Bluetooth: Audio: CAP implement broadcast reception stop

kruithofa opened the pull request:
Implement the CAP broadcast reception stop procedures and add unit tests

Commit:
Bluetooth: Audio: CAP implement broadcast reception stop

Implement the CAP broadcast reception stop procedures and
add unit tests

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/audio/cap.h:
```diff
@@ -945,14 +956,26 @@ int bt_cap_commander_broadcast_reception_start(
+	/** Source ID of the receive state. */
+	uint8_t src_id;
+
+	/** Number of subgroups */
+	size_t num_subgroups;
```
I think the description of this needs improvement. 

IIRC this value shall be the reported number of subgroups in the specified receive state, right? (At least until we may see a fix in a future version of BASS).

That should be added here.

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+	uint16_t total_bis_sync;
+	uint8_t src_id = 0;
+
+	if (err != 0) {
+		LOG_DBG("BASS recv state read failed (%d)\n", err);
+		return;
+	}
```
We are never reading the receive state, so this cannot happen.

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+		return;
+	}
+
+	if (state == NULL) {
+		/* Empty receive state */
+		return;
+	}
```
Suggest to add a comment why we ignore this

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+		return;
+	}
+
+	bt_addr_le_to_str(&state->addr, le_addr, sizeof(le_addr));
+
+	LOG_DBG("BASS recv state: src_id %u, addr %s, sid %u, sync_state %u", state->src_id,
+		le_addr, state->adv_sid, state->pa_sync_state);
```
Not point in logging before `bt_cap_common_conn_in_active_proc(conn)` IMO (and suggest to only log the conn pointer rather than the address).

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+	LOG_DBG("BASS recv state: src_id %u, addr %s, sid %u, sync_state %u", state->src_id,
+		le_addr, state->adv_sid, state->pa_sync_state);
+
+	if (state->encrypt_state == BT_BAP_BIG_ENC_STATE_BAD_CODE) {
+		LOG_DBG("Encryption state is BT_BAP_BIG_ENC_STATE_BAD_CODE");
+		return;
+	}
```
Why do we care about this?

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+	struct bt_conn *conn, int err, const struct bt_bap_scan_delegator_recv_state *state)
+{
+	struct bt_cap_common_proc *active_proc = bt_cap_common_get_active_proc();
+	struct bt_conn *cap_conn = NULL;
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	uint16_t total_bis_sync;
+	uint8_t src_id = 0;
```
Looks like several of these can be moved inside the `if (bt_cap_common_conn_in_active_proc(conn)` clause. Alternatively, you can do as we do in the CAP initiator with
```c
	if (!bt_cap_common_conn_in_active_proc(conn)) {

		/* State change happened outside of a procedure; ignore */
		return;
	}
```

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+
+			total_bis_sync |= subgroup->bis_sync;
+		}
+		for (size_t i = 0; i < active_proc->proc_cnt; i++) {
+			if (state->src_id == active_proc->proc_param.commander[i]
+						     .broadcast_reception_stop.src_id &&
+			    conn == active_proc->proc_param.commander[i].conn) {
+				src_id = state->src_id;
+				cap_conn = conn;
+			}
+		}
```
What's the purpose of this loop?

Won't `src_id` always be `state->src_id` and `cap_conn` always be `conn`?

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+	if (bt_cap_common_conn_in_active_proc(conn) &&
+	    active_proc->proc_type == BT_CAP_COMMON_PROC_TYPE_BROADCAST_RECEPTION_STOP) {
+		total_bis_sync = 0;
+		for (uint8_t i = 0; i < state->num_subgroups; i++) {
+			const struct bt_bap_bass_subgroup *subgroup = &state->subgroups[i];
+
+			total_bis_sync |= subgroup->bis_sync;
+		}
```
You can `break` after the first non-0 value for early termination

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+			}
+		}
+
+		if (total_bis_sync == 0 && cap_conn == conn) {
```
What's the purpose of the `cap_conn == conn` comparison?

kruithofa replied:
I rewrote this part now

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+				(void *)conn, err);
+			bt_cap_common_abort_proc(conn, err);
+			cap_commander_proc_complete();
+		}
+
+	} else {
```
```suggestion
		}
	} else {
```

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+			LOG_DBG("Removing source for conn %p", (void *)cap_conn);
+			err = bt_bap_broadcast_assistant_rem_src(cap_conn, src_id);
+			if (err != 0) {
+				LOG_DBG("Failed to perform broadcast reception stop for conn %p: "
+					"%d",
```
```suggestion
				LOG_DBG("Failed to rem_src for conn %p: %d",
```

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+		active_proc->proc_initiated_cnt++;
+		err = bt_bap_broadcast_assistant_mod_src(conn, &mod_src_param);
+		if (err != 0) {
+			LOG_DBG("Failed to perform broadcast reception stop for conn %p: %d",
+				(void *)conn, err);
```
```suggestion
			LOG_DBG("Failed to mod_src for conn %p: %d", (void *)conn, err);
```

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+	if (!bt_cap_common_proc_is_done()) {
+		struct bt_cap_commander_proc_param *proc_param;
+
+		proc_param = &active_proc->proc_param.commander[active_proc->proc_done_cnt];
```
Is this resilient to server-induced remove source? 

For example if we are removing the source for scan delegators SD_1, SD_2 and SD_3, and we just finished removing for SD_1 and started removing for SD_2 but SD_3 removed the source by itself, then then we increment `proc_done_cnt` here then there is no remove state to remove. 

We had a similar issue for the CAP Initiator until we implemented the `update_proc_done_cnt` function, which does state checks for each procedure rather than a simple counter, which allows it to better do the right thing if the remote devices are modified during the procedure

Thalley replied:
ping @kruithofa

kruithofa replied:
as discussed, issue #78166 has been created

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+			return false;
+		}
+
+		for (size_t j = 0U; j < i; j++) {
+			const union bt_cap_set_member *other = &param->param[j].member;
+
+			if (other == member) {
+				LOG_DBG("param->members[%zu] (%p) is duplicated by "
+					"param->members[%zu] (%p)",
+					j, other, i, member);
+				return false;
+			}
+		}
```
If we want to apply a duplication check, then it needs to be duplicate of `{conn, src_id}`, right? There shouldn't be anything wrong with stopping multiple receive states on the same connection, right?

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+
+	err = bt_bap_broadcast_assistant_mod_src(conn, &mod_src_param);
+	if (err != 0) {
+		LOG_DBG("Failed to stop broadcast reception for conn %p: %d", (void *)conn, err);
+
+		return -ENOEXEC;
```
Should we just always return `-ENOEXEC` here? Could we check for known/expected errors and send them to the application?

kruithofa replied:
we already check for invalid parameters, or if there is already a CAP procedure running

Thalley replied:
My point is that if `bt_bap_broadcast_assistant_mod_src` returns e.g. `-EBUSY` or `ENOTCONN`, isn't it better to send those the application and document these as possible return values? 

For example if it returns `-EBUSY`, then the application can simply retry with a `k_work` or something.

Thalley commented on subsys/bluetooth/audio/cap_internal.h:
```diff
@@ -92,6 +93,12 @@ struct cap_broadcast_reception_start {
+struct cap_broadcast_reception_stop {
+	uint8_t src_id;
+	uint8_t num_subgroups;
+	struct bt_bap_bass_subgroup subgroups[CONFIG_BT_BAP_BASS_MAX_SUBGROUPS];
```
Do we need to store this statically? The subgroups for the remove source could probably just be stack-allocated as needed, right? Or would that require too much stack memory?

kruithofa replied:
It could be stack allocated, but we wouldn't gain anything. The `struct cap_broadcast_reception_stop` is used in the `struct bt_cap_commander_proc_param` as part of a union

```
#if defined(CONFIG_BT_BAP_BROADCAST_ASSISTANT)
		struct cap_broadcast_reception_start broadcast_reception_start;
		struct cap_broadcast_reception_stop broadcast_reception_stop;
#endif /* CONFIG_BT_BAP_BROADCAST_ASSISTANT */
```

so there wouldn't be any memory savings

Thalley commented on tests/bluetooth/audio/cap_commander/uut/bap_broadcast_assistant.c:
```diff
@@ -28,3 +28,40 @@ int bt_bap_broadcast_assistant_add_src(struct bt_conn *conn,
+
+	if (broadcast_assistant_cbs->mod_src != NULL) {
+		broadcast_assistant_cbs->mod_src(conn, 0);
+	}
+	if (broadcast_assistant_cbs->recv_state != NULL) {
```
```suggestion
	}
	
	if (broadcast_assistant_cbs->recv_state != NULL) {
```

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +402,162 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+	err = bt_cap_commander_broadcast_reception_stop(&fixture->stop_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+
+	zexpect_call_count("bt_cap_commander_cb.broadcast_reception_stop", 1,
+			   mock_cap_commander_broadcast_reception_stop_cb_fake.call_count);
```
Consider also validating the parameters in the callback

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +402,162 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+			   mock_cap_commander_broadcast_reception_stop_cb_fake.call_count);
+}
+
+ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_stop_double)
+{
+	int err;
+
+	err = bt_cap_commander_register_cb(&mock_cap_commander_cb);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+	printk("Source ID: %d\n", fixture->stop_param.param[0].src_id);
+	err = bt_cap_commander_broadcast_reception_stop(&fixture->stop_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+
+	err = bt_cap_commander_broadcast_reception_stop(&fixture->stop_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+
+	zexpect_call_count("bt_cap_commander_cb.broadcast_reception_stop", 2,
+			   mock_cap_commander_broadcast_reception_stop_cb_fake.call_count);
+}
```
This is a kind of weird test. 
In an actual situation the first `reception_stop` would remove the receive state, and thus the second one would always fail to do so. It would, however, fail with an error in the callback.

Thalley commented on tests/bluetooth/audio/cap_commander/uut/bap_broadcast_assistant.c:
```diff
@@ -28,3 +28,40 @@ int bt_bap_broadcast_assistant_add_src(struct bt_conn *conn,
+		state.subgroups[i].bis_sync = param->subgroups[i].bis_sync;
+	}
+
+	if (broadcast_assistant_cbs->mod_src != NULL) {
+		broadcast_assistant_cbs->mod_src(conn, 0);
+	}
```
Consider adding a check to whether there's a receive state or not here. Once `bt_bap_broadcast_assistant_rem_src` have been called, `bt_bap_broadcast_assistant_mod_src` should fail with the same `{conn, src_id}`

Thalley reviewed (changes requested):
Are the BSIM tests passing with this implementation? 

I might be missing something, but looks like reception stop might not work as intended on multiple connections

kruithofa commented:
> Are the BSIM tests passing with this implementation?
> 
> I might be missing something, but looks like reception stop might not work as intended on multiple connections

In the broadcast_reception_stop procedure we store the parameters, including the connections on which to act. How often we need to perform a mod_src/rem_src operation is also stored by calling bt_cap_common_start_proc with param->count

Commit:
Bluetooth: Audio: improve broadcast reception stop unittest

Add more parameter checking for the unittests for the
CAP procedure 'broadcast reception stop'

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,280 @@ int bt_cap_commander_broadcast_reception_start(
+{
+	struct bt_cap_common_proc *active_proc = bt_cap_common_get_active_proc();
+
+	if (state == NULL) {
+		/* Empty receive state, so there is no BIS information available which
+		 * can trigger a remove source operation
+		 */
+		return;
+	}
```
If the `state` is `NULL` it indicates that the source was removed

Thalley commented on subsys/bluetooth/audio/cap_commander.c:
```diff
@@ -399,14 +400,280 @@ int bt_cap_commander_broadcast_reception_start(
+
+	if (bt_cap_common_conn_in_active_proc(conn) &&
+	    active_proc->proc_type == BT_CAP_COMMON_PROC_TYPE_BROADCAST_RECEPTION_STOP) {
+		bool bis_sync_zero = true;
+
+		LOG_DBG("BASS recv state: conn %p, src_id %u", (void *)conn, state->src_id);
+
+		for (uint8_t i = 0; i < state->num_subgroups; i++) {
+			const struct bt_bap_bass_subgroup *subgroup = &state->subgroups[i];
+
+			if (subgroup->bis_sync != 0) {
+				bis_sync_zero = false;
+				break;
+			}
```
You could actually remove the `bool` and just `return` instead

Thalley replied:
```suggestion
			if (subgroup->bis_sync != 0) {
				/* Receive state not stopped yet, wait for another notification */
				bis_sync_zero = false;
				break;
			}
```

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -32,21 +32,43 @@ LOG_MODULE_REGISTER(bt_broadcast_reception_test, CONFIG_BT_CAP_COMMANDER_LOG_LEV
+	struct bt_cap_commander_broadcast_reception_stop_param stop_param;
 };
 
+/* src_id can not be part of the fixture since it is accessed in the callback function */
+static uint8_t src_id[CONFIG_BT_MAX_CONN];
+static struct bt_bap_broadcast_assistant_cb broadcast_assistant_cb;
```
Ideally we wouldn't have global variables, but I never found a way to access the current test fixture for async functionality like this. 

We do, however, need to clear `src_id` between each test. 

`broadcast_assistant_cb` could/should be moved to the fixture

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -32,21 +32,43 @@ LOG_MODULE_REGISTER(bt_broadcast_reception_test, CONFIG_BT_CAP_COMMANDER_LOG_LEV
+	test_stop_param_init(fixture);
+
+	broadcast_assistant_cb.recv_state = cap_commander_broadcast_assistant_recv_state_cb;
+	bt_bap_broadcast_assistant_register_cb(&broadcast_assistant_cb);
```
Should check the return value of `bt_bap_broadcast_assistant_register_cb`. I imagine it will fail on the second test as you never unregister the callbacks, and registering the same callback structure twice will return an error.

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -91,9 +113,9 @@ static void test_start_param_init(void *f)
-		fixture->start_subgroups[i].bis_sync = i;
-		fixture->start_subgroups[i].metadata_len = 0;
+	for (size_t i = 0; i < ARRAY_SIZE(fixture->subgroups); i++) {
+		fixture->subgroups[i].bis_sync = 1 << i;
```
Why are we shifting the bis_sync?

Thalley replied:
ping @kruithofa

kruithofa replied:
To ensure that each subgroup has a different bis_sync

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -111,9 +133,27 @@ static void test_start_param_init(void *f)
+	int err;
+
+	fixture->stop_param.type = BT_CAP_SET_TYPE_AD_HOC;
+	fixture->stop_param.param = fixture->stop_member_params;
+
+	fixture->stop_param.count = ARRAY_SIZE(fixture->stop_member_params);
```
```suggestion
	fixture->stop_param.param = fixture->stop_member_params;
	fixture->stop_param.count = ARRAY_SIZE(fixture->stop_member_params);
```

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -111,9 +133,27 @@ static void test_start_param_init(void *f)
+		fixture->stop_member_params[i].num_subgroups = CONFIG_BT_BAP_BASS_MAX_SUBGROUPS;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(fixture->conns); i++) {
+		err = bt_cap_commander_discover(&fixture->conns[i]);
+		zassert_equal(0, err, "Unexpected return value %d", err);
```
This does not seem to fit in a function called `test_stop_param_init`

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +411,220 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+	err = bt_cap_commander_broadcast_reception_start(&fixture->start_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+
+	zexpect_call_count("bt_cap_commander_cb.broadcast_reception_start", 1,
+			   mock_cap_commander_broadcast_reception_start_cb_fake.call_count);
+	zassert_equal_ptr(NULL,
+			  mock_cap_commander_broadcast_reception_start_cb_fake.arg0_history[0]);
+	zassert_equal(0, mock_cap_commander_broadcast_reception_start_cb_fake.arg1_history[0]);
```
I wonder if we can make a helper function for these 3 checks. Collectively they check if the callback is called once and without error, right?

Thalley replied:
Same can be applied to the reception stop callback

kruithofa replied:
Yes, I created a follow-up PR for that

Thalley replied:
Any reason not to do it now and here in this PR?

kruithofa replied:
I want to use the same helper function for the broadcast_reception_start tests, but I could do it as a separate commit

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +411,220 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+
+	zexpect_call_count("bt_cap_commander_cb.broadcast_reception_start", 1,
+			   mock_cap_commander_broadcast_reception_start_cb_fake.call_count);
+	zassert_equal_ptr(NULL,
+			  mock_cap_commander_broadcast_reception_start_cb_fake.arg0_history[0]);
```
Shouldn't this never be NULL?

kruithofa replied:
in 'cap_commander_complete', where the callback functions are called, the parameter 0 is 'failed_conn', which is NULL in case there was no error

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +411,220 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+	err = bt_cap_commander_broadcast_reception_stop(&fixture->stop_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
+
+	err = bt_cap_commander_broadcast_reception_stop(&fixture->stop_param);
+	zassert_equal(0, err, "Unexpected return value %d", err);
```
Should the second stop return something like `-EALREADY` or `-ENOEXEC` as it doesn't do anything? It is OK that it returns `0`

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +411,220 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+	zexpect_call_count("bt_cap_commander_cb.broadcast_reception_stop", 1,
+			   mock_cap_commander_broadcast_reception_stop_cb_fake.call_count);
+
+	/*
+	 * Since the 2nd reception_stop procedure never completed initiating a new CAP procedure
+	 * should result in an error -EBUSY
+	 */
+	err = bt_cap_commander_broadcast_reception_start(&fixture->start_param);
+	zassert_equal(-EBUSY, err, "Unexpected return value %d", err);
```
This does not make sense to me. 

A call to `bt_cap_commander_broadcast_reception_stop` that doesn't do anything because they are all stopped should not make a call to `bt_cap_commander_broadcast_reception_start` return an error.

kruithofa replied:
Similar to above, in the unittest the stop procedure never finishes, so a 2nd start will return that there is already a procedure running.

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +411,220 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+	err = bt_cap_commander_broadcast_reception_start(&fixture->start_param);
+	zassert_equal(-EBUSY, err, "Unexpected return value %d", err);
+
+	/*
+	 * We need to cleanup since the CAP commander remembers state, otherwise the next test will
+	 * fail.
+	 * Since this cleanup is not normally necessary we do it here and not in
+	 * cap_commander_test_broadcast_reception_after
+	 */
+	err = bt_cap_commander_cancel();
+	zassert_equal(0, err, "Unexpected return value: %d", err);
+}
```
Consider always calling `cancel` in the cleanup function of the test suite. If any tests fails for any reason, then it will be left in a `busy` state. 

We should always attempt to clean up everything after any test so that if Test A fails it does not affect Test B

Thalley commented on tests/bluetooth/audio/cap_commander/uut/bap_broadcast_assistant.c:
```diff
@@ -6,24 +6,160 @@
+	zassert_not_null(inst, "inst is NULL");
+
+	max_src_id++;
+	inst->recv_states[0].src_id = max_src_id;
+	inst->recv_states[0].past_avail = false;
+	inst->recv_states[0].adv_sid = param->adv_sid;
+	inst->recv_states[0].broadcast_id = param->broadcast_id;
+	inst->inst_active = true;
+	state.pa_sync_state = param->pa_sync;
+	state.src_id = max_src_id;
+	state.num_subgroups = param->num_subgroups;
+	for (size_t i = 0; i < param->num_subgroups; i++) {
+		state.subgroups[i].bis_sync = param->subgroups[i].bis_sync;
+	}
+
+	bt_addr_le_copy(&inst->recv_states[0].addr, &param->addr);
```
We should either access the correct `recv_states` instead of always `0`, or not have it as an array.

Thalley commented on tests/bluetooth/audio/cap_commander/uut/bap_broadcast_assistant.c:
```diff
@@ -6,24 +6,160 @@
+
+	zassert_not_null(conn, "conn is NULL");
+
+	inst = inst_by_conn(conn);
+	zassert_true(inst->inst_active, "Instant is not active");
+	zassert_equal(src_id, inst->recv_states[0].src_id, "Invalid src_id");
+
```
Should we check for the `pa_sync` state and `bis_sync` state as well?

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -61,16 +89,31 @@ static void *cap_commander_test_broadcast_reception_setup(void)
+	bt_bap_broadcast_assistant_unregister_cb(&fixture->broadcast_assistant_cb);
 
+	/* We need to cleanup since the CAP commander remembers state */
+	if (bt_cap_common_proc_is_active()) {
+		err = bt_cap_commander_cancel();
+		zassert_equal(0, err, "Unexpected return value: %d", err);
+	}
```
```suggestion
	bt_cap_commander_cancel();
```

We can just always attempt to cancel and ignore any errors (or ignore `-EALREADY`). We don't need to hack `bt_cap_common_proc_is_active` to check whether it's active or not.

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -371,3 +450,172 @@ ZTEST_F(cap_commander_test_broadcast_reception, test_commander_reception_start_i
+
+	test_broadcast_reception_stop(&fixture->stop_param);
+
+	/*
+	 * Since the 2nd reception_stop procedure never completed initiating a new CAP procedure
+	 * should result in an error -EBUSY
+	 * we try with broadcast_reception_start, but could use any CAP procedure for this test
+	 */
+	err = bt_cap_commander_broadcast_reception_start(&fixture->start_param);
+	zassert_equal(-EBUSY, err, "Unexpected return value %d", err);
```
Continue from: https://github.com/zephyrproject-rtos/zephyr/pull/76263#discussion_r1715310725

Why doesn't the second `test_broadcast_reception_stop` never complete? It should either complete or return an error. Getting stuck in a busy state seems incorrect

kruithofa replied:
The second test_broadcast_reception_stop never completes because we do not get a receive-state callback. However, in the broadcast_reception_stop procedure we can not know if we will never get a callback, since we do not know if someone else started broadcast reception or not, i.e. remembering whether we already did a broadcast reception stop, or not, is not going to fix this issue. We could add something like a time-out, but as per previous discussion this is better left to be handled in the application

Thalley replied:
Since we always do a GATT write with response, we will _always_ (unless the remote server is non-compliant) get a response from our write, and as such there will always be a callback. 

For example if the receive state already has pa_sync_state = 0 and BIS_sync state = 0 (not synced to either), then the mod source operation will likely not trigger a receive state change, but may still provide a write response with `BT_ATT_ERR_SUCCESS`, which should trigger the next step of the stop (rem_src). 

If on the other hand the source is already removed, then we would get a write response with `BT_BAP_BASS_ERR_INVALID_SRC_ID` which we can use to trigger the completed callback with the error. 

Do you have an example where we cannot decidedly complete the CAP procedure?
